twenty-first/src/math/polynomial.rs

```diff
@@ -1360,6 +1409,388 @@ where
+            Self::batch_fast_coset_extrapolate(domain_offset, codeword_length, codewords, points)
+        } else {
+            Self::batch_naive_coset_extrapolate(domain_offset, codeword_length, codewords, points)
+        }
+    }
+
+    fn batch_fast_coset_extrapolate(
+        domain_offset: BFieldElement,
+        codeword_length: usize,
+        codewords: &[FF],
+        points: &[FF],
+    ) -> Vec<FF> {
+        let n = codeword_length;
+
+        let zerofier_tree = ZerofierTree::new_from_domain(points);
+        let modulus = zerofier_tree.zerofier();
+        let (even_zerofiers, odd_zerofiers, shift_coefficients, tail_length) =
+            Self::fast_modular_coset_interpolate_preprocess(
+                codeword_length,
+                domain_offset,
+                &modulus,
+            );
+
+        (0..codewords.len() / n)
+            .flat_map(|i| {
+                let codeword = &codewords[i * n..(i + 1) * n];
+                let minimal_interpolant =
+                    Self::fast_modular_coset_interpolate_with_zerofiers_and_ntt_friendly_multiple(
+                        codeword,
+                        domain_offset,
+                        &modulus,
+                        &even_zerofiers,
+                        &odd_zerofiers,
+                        &shift_coefficients,
+                        tail_length,
+                    );
+                minimal_interpolant.divide_and_conquer_batch_evaluate(&zerofier_tree)
+            })
+            .collect()
+    }
+
+    fn batch_naive_coset_extrapolate(
+        domain_offset: BFieldElement,
+        codeword_length: usize,
+        codewords: &[FF],
+        points: &[FF],
+    ) -> Vec<FF> {
+        let (zerofier_tree, shift_coefficients, tail_length) =
+            Self::naive_coset_extrapolate_preprocessing(points);
+        let n = codeword_length;
+        let logn = n.ilog2();
+
+        (0..codewords.len() / n)
+            .flat_map(|i| {
+                let mut coefficients = codewords[i * n..(i + 1) * n].to_vec();
+                intt(
+                    &mut coefficients,
+                    BFieldElement::primitive_root_of_unity(n as u64).unwrap(),
+                    logn,
+                );
+                Polynomial::new(coefficients)
+                    .scale(FF::from(domain_offset.inverse().value()))
+                    .reduce_by_ntt_friendly_modulus(&shift_coefficients, tail_length)
+                    .divide_and_conquer_batch_evaluate(&zerofier_tree)
+            })
+            .collect()
+    }
+
+    /// Parallel version of [`batch_coset_extrapolate`](Self::batch_coset_extrapolate).
+    pub fn par_batch_coset_extrapolate(
+        domain_offset: BFieldElement,
+        codeword_length: usize,
+        codewords: &[FF],
+        points: &[FF],
+    ) -> Vec<FF> {
+        if points.len() < Self::FAST_COSET_EXTRAPOLATE_THRESHOLD {
+            Self::par_batch_fast_coset_extrapolate(
```

Comment:
The direction of the inequality sign here is consistent with that in `coset_extrapolate`, which is the one that is benchmarked. Switching the direction there gives rise (as per benchmark data) to a dispatcher that consistently picks the slower method.

The reason is that `fast_coset_extrapolate` works betst when the number of extrapolation points is small (relative to the codeword domain). Otherwise you might as well use the alternative, which is INTT followed by mod-reduce and then batch-evaluate.

You are welcome to suggest better names for the alternative code path.